File: src/main/java/de/tanschmi/aoc2024/dec2/Dec2.java
package de.tanschmi.aoc2024.dec2;

import de.tanschmi.aoc2024.ListReader;
import lombok.extern.slf4j.Slf4j;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

@Slf4j
public class Dec2 {

    public int step1(File file) throws IOException {

        ListReader reader = new ListReader();
        List<Sequence> sequences = reader.readSequences(file);

        int sum = countSafe(sequences);
        log.info("Summe: {}", sum);
        return sum;
    }


    private int countSafe(List<Sequence> sequences) {
        int sum = 0;
        for (Sequence seq : sequences) {
            boolean safe = isSequenceSafe(seq.getColumns());
            if (safe) {
                sum++;
            }
        }
        return sum;
    }

    boolean isSequenceSafe(ArrayList<Integer> columns) {

        Boolean increasing = null;
        for (int i = 1; i < columns.size(); i++) {
            int previous = columns.get(i - 1);
            int current = columns.get(i);
            if (increasing == null) {
                increasing = previous > current;
            } else {
                //Wert würde sich ändern -> unsafe
                if (increasing != previous > current) {
                    return false;
                }
            }
            if (previous > current) {
                int diff = previous - current;
                if (diff > 3 || diff < 1) {
                    return false;
                }
            } else {
                int diff = current - previous;
                if (diff > 3 || diff < 1) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: src/main/kotlin/de/tanschmi/kotlin/aoc2024/dec2/Dec2.kt
package de.tanschmi.kotlin.aoc2024.de.tanschmi.kotlin.aoc2024.dec2

import de.tanschmi.kotlin.FileReader
import org.apache.commons.io.FileUtils
import org.apache.logging.log4j.kotlin.logger
import java.io.File
import java.nio.charset.Charset

const val LIMIT = 3;

class Task1 {
    companion object {
        private val log = logger()
    }

    fun task1(input: File): Int {
        var sequences: ArrayList<Sequence> = readSequences(input)
        var sum = countSafeSequences(sequences)
        log.info("Summe für Task 1: $sum")
        return sum
    }

    fun countSafeSequences(sequences: ArrayList<Sequence>): Int {
        var sum: Int = 0
        for (sequence in sequences) {
            var isSafe: Boolean = isSafeSequence(sequence)
            if (isSafe) {
                sum++;
            }
        }
        return sum;
    }

    fun isSafeSequence(sequence: Sequence): Boolean {
        var increasing: Boolean? = null;
        for (i in 1..(sequence.numbers.size - 1)) {
            var first = sequence.numbers.get(i - 1)
            var secnd = sequence.numbers.get(i)
            if (first > secnd) {
                if (increasing == true) {
                    return false
                };
                if (increasing == null) {
                    increasing = false
                }
                if (first - secnd > LIMIT || first - secnd <= 0) {
                    return false;
                }
            } else {
                if (increasing == false) {
                    return false;
                }
                if (increasing == null) {
                    increasing = true
                }
                if (secnd - first > LIMIT || secnd - first <= 0) {
                    return false;
                }
            }
        }
        return true;
    }

    fun readSequences(file: File): ArrayList<Sequence> {
        var lines: List<String> = FileUtils.readLines(file, Charset.defaultCharset())
        var sequences: ArrayList<Sequence> = ArrayList<Sequence>()

        for (line in lines) {
            var splitted = line.split(" ")
            var seq: Sequence = Sequence()
            for (s in splitted) {
                seq.numbers.add(s.toInt())
            }
            sequences.add(seq);
        }
        return sequences
    }
}

class Task2 {
    companion object {
        private val log = logger()
    }

    fun task2(input: File): Int {

        var sequences: ArrayList<Sequence> = Task1().readSequences(input)
        var sum = countSafeSequences(sequences)
        log.info("Summe für Task 2: $sum")
        return sum
    }

    private fun countSafeSequences(sequences: ArrayList<Sequence>): Int {
        var sum: Int = 0
        for (sequence in sequences) {
            var isSafe: Boolean = isSafeSequence(sequence)
            if (isSafe) {
                sum++;
            }
        }
        return sum;
    }

    fun isSafeSequence(sequence: Sequence): Boolean {
        var task1 = Task1()


        val subsequences = sequence.getSubsequences()
        for (s in subsequences) {
            if (task1.isSafeSequence(s)) {
                return true
            }
        }
        return false;
        //return task1.isSafeSequence(sequence);
    }


}

class Sequence {
    var numbers: ArrayList<Int> = ArrayList<Int>();

    fun getSubsequences(): List<Sequence> {
        var list = ArrayList<Sequence>()
        for (i in 0..(numbers.size - 1)) {
            var newlist = ArrayList<Int>(numbers.filterIndexed { index, _ -> index != i })
            var newSeq = Sequence()
            newSeq.numbers = newlist
            list.add(newSeq)
        }
        return list
    }


}